shell/platform/darwin/macos/framework/Source/FlutterMenuPlugin.mm

```diff
@@ -0,0 +1,418 @@
+  NSArray<NSMenuItem*>* _platformProvidedItems;
+  // These are the menu delegates that will listen to open/close events for menu
+  // items. This array is holding them so that we can deallocate them when
+  // rebuilding the menus.
+  NSMutableArray<FlutterMenuDelegate*>* _menuDelegates;
+}
+
+#pragma mark - Private Methods
+
+- (instancetype)initWithChannel:(FlutterMethodChannel*)channel {
+  self = [super init];
+  if (self) {
+    _channel = channel;
+    _platformProvidedItems = @[];
+    _menuDelegates = [[NSMutableArray alloc] init];
+
+    // Make a copy of all the platform provided menus for later use.
+    _platformProvidedItems = [[NSApp.mainMenu itemArray] mutableCopy];
+
+    // As copied, these platform provided menu items don't yet have the APP_NAME
+    // string replaced in them, so this rectifies that.
+    [self replaceAppName:_platformProvidedItems];
+  }
+  return self;
+}
+
+/**
+ * Iterates through the given menu hierarchy, and replaces "APP_NAME"
+ * with the localized running application name.
+ */
+- (void)replaceAppName:(NSArray<NSMenuItem*>*)items {
+  NSString* appName = [NSRunningApplication currentApplication].localizedName;
+  for (NSMenuItem* item in items) {
+    if ([[item title] containsString:kAppName]) {
+      [item setTitle:[[item title] stringByReplacingOccurrencesOfString:kAppName
+                                                             withString:appName]];
+    }
+    if ([item hasSubmenu]) {
+      [self replaceAppName:[[item submenu] itemArray]];
+    }
+  }
+}
+
+- (NSMenuItem*)findProvidedMenu:(NSMenu*)menu ofType:(SEL)selector withTag:(int)tag {
+  const NSArray<NSMenuItem*>* items = menu != nil ? menu.itemArray : _platformProvidedItems;
+  for (NSMenuItem* item in items) {
+    if ([item action] == selector && (tag == 0 || [item tag] == tag)) {
+      return item;
+    }
+    if ([[item submenu] numberOfItems] > 0) {
+      NSMenuItem* foundChild = [self findProvidedMenu:[item submenu] ofType:selector withTag:tag];
+      if (foundChild != nil) {
+        return foundChild;
+      }
+    }
+  }
+  return nil;
+}
+
+- (NSMenuItem*)createPlatformProvidedMenu:(PlatformProvidedMenu)type {
+  const std::map<int, SEL> providedMenus = GetMacOSProvidedMenus();
+  auto found_type = providedMenus.find(type);
+  if (found_type == providedMenus.end()) {
+    return nil;
+  }
+  SEL selectorTarget = found_type->second;
+  // Since it doesn't have a definitive selector, the Services submenu is
+  // assumed to be the first item with a submenu action in the first menu item
+  // of the default menu set. We can't just get the title to check, since that
+  // is localized, and the contents of the menu aren't fixed (or even available).
+  NSMenu* startingMenu =
+      type == PlatformProvidedMenu::servicesSubmenu ? [_platformProvidedItems[0] submenu] : nil;
+  NSMenuItem* found = [self findProvidedMenu:startingMenu ofType:selectorTarget withTag:0];
```

Comment:
Is the reason for `withTag:0` here that we _expect_ tag 0? Or that we don't care about the tag, just the selector, since we special-case tag 0 in `findProvidedMenu:ofType:withTag:`? If the latter, is the special-casing actually necessary?

Reply:
It meant that we didn't care, but since we only ever called it with zero, I just nuked the `withTag` argument.